test/src/main/java/org/apache/accumulo/test/functional/ZooMutatorIT.java

```diff
@@ -0,0 +1,139 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.test.functional;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+import org.apache.accumulo.core.client.Accumulo;
+import org.apache.accumulo.core.clientImpl.ClientContext;
+import org.apache.accumulo.core.conf.Property;
+import org.apache.accumulo.fate.zookeeper.ZooReaderWriter;
+import org.apache.accumulo.harness.AccumuloClusterHarness;
+import org.apache.accumulo.test.categories.MiniClusterOnlyTests;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+import com.google.common.io.BaseEncoding;
+
+@Category(MiniClusterOnlyTests.class)
+public class ZooMutatorIT extends AccumuloClusterHarness {
+  /**
+   * A simple stress test that looks for race conditions in
+   * {@link ZooReaderWriter#mutateOrCreate(String, byte[], org.apache.accumulo.fate.zookeeper.ZooReaderWriter.Mutator)}
+   */
+  @Test
+  public void concurrentMutatorTest() throws Exception {
+    try (var client = Accumulo.newClient().from(getClientProps()).build();
+        var context = (ClientContext) client) {
+      String secret = cluster.getSiteConfiguration().get(Property.INSTANCE_SECRET);
+
+      ZooReaderWriter zk = new ZooReaderWriter(context.getZooKeepers(),
+          context.getZooKeepersSessionTimeOut(), secret);
+
+      var executor = Executors.newFixedThreadPool(16);
+
+      String initialData = hash("Accumulo Zookeeper Mutator test data") + " 0";
+
+      List<Future<?>> futures = new ArrayList<>();
+
+      // This map is used to ensure multiple threads do not successfully write the same value and no
+      // values are skipped. The hash in the value also verifies similar things in a different way.
+      ConcurrentHashMap<Integer,Integer> countCounts = new ConcurrentHashMap<>();
+
+      for (int i = 0; i < 16; i++) {
+        futures.add(executor.submit(() -> {
+          try {
+
+            int count = -1;
+            while (count < 200) {
+              byte[] val =
+                  zk.mutateOrCreate("/test-zm", initialData.getBytes(UTF_8), this::nextValue);
+              int nextCount = getCount(val);
+              assertTrue("nextCount <= count " + nextCount + " " + count, nextCount > count);
+              count = nextCount;
+              countCounts.merge(count, 1, Integer::sum);
+            }
+
+          } catch (Exception e) {
+            throw new RuntimeException(e);
+          }
+        }));
+      }
+
+      // wait and check for errors in background threads
+      for (Future<?> future : futures) {
+        future.get();
+      }
```

Comment:
You could probably get a one-liner out of this:

```suggestion
      futures.forEach(f -> f.get());
```

I'm still uncertain about how the Futures handle RTEs and/or AssertionErrors. Will those be caught and preserved in the Future object for later, when `.get()` is called?

Reply:
Any exception occurring in the background thread will be wrapped in a ExecutionException and thrown to the foreground thread when calling get(), so the RTEs and AssertionErrors in the background would show up that way. The one liner probably would not work because get() throws ExecutionException

Reply:
Ah, okay, that makes sense. Thanks for the explanation.